exchanges/graphcache/src/operations/query.ts

```diff
@@ -529,6 +532,7 @@ const readSelection = (
       !deferRef &&
       dataFieldValue === undefined &&
       (directives.optional ||
+        (optionalRef && !directives.required) ||
```

Comment:
This doesn't seem right.
We're of course not tracking `optionalRef` per field, but per “selection iteration”, which is fine but `directives.required` here isn't the only usage.

We also have `directives.required` below on `dataFieldValue === null`, so that error isn't cascading:
https://github.com/urql-graphql/urql/blob/153a250bd5a0ac7eb0b73397a3f5452eb2c2cb93/exchanges/graphcache/src/operations/query.ts#L543-L556

It should probably be `(directives.required || optionalRef === false)` rather than `directives.required` to also cover that case.

Leaving the comment here, since I can't comment on the unmodified line

Reply:
`optionalRef` is false by default so changing that line will have a much broader effect than what we are aiming for I reckon 😅 we explicitly check whether this has a required directive and else allow ourselves to fall through to the default `optional = false` case as we did before.